### PR TITLE
refactor: extract localized logic from Upload

### PR DESCRIPTION
**Pull request by benjycui**
ref: https://github.com/ant-design/ant-design/issues/5103

### CODE REVIEW

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -40,11 +35,7 @@ export default class Upload extends React.Component<UploadProps, any> {
     supportServerRender: true,
   };
 
-  static contextTypes = {
-    antLocale: React.PropTypes.object,
-  };
-
-  context: UploadContext;
+  static contextTypes = UploadLocalizable.contextTypes;
```
Comment (benjycui) [inline components/upload/Upload.tsx]: To solve issues which similar to: https://github.com/ant-design/ant-design/issues/5484

**Comment by benjycui** [comment]
This solution works fine with TypeScript and IE<=10

If you think it OK, I will replace `injectLocale` with this solution.

File: components/locale-provider/Localizable.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export interface ComponentContext {
+  antLocale?: { [key: string]: any };
+}
+
+export default function<P, S>(componentName: string, defaultLocale) {
+  return class Localizable extends React.Component<P & ComponentProps, S> {
+    static contextTypes = {
+        antLocale: PropTypes.object,
```
Comment (afc163) [inline components/locale-provider/Localizable.tsx]: indent

File: components/upload/Upload.tsx (hunk 1 above)
Comment (afc163) [inline components/upload/Upload.tsx]: Localizable is used only in Upload, or it will be replace `injectLocale`?

Comment (benjycui) [inline components/upload/Upload.tsx]: > If you think it OK, I will replace injectLocale with this solution.

Comment (bang88) [inline components/upload/Upload.tsx]: Here need consider generic Component. like Table.

**Review by BANG88 — COMMENTED** [review]
These changes LGTM.

**Comment by benjycui** [comment]
There is no TypeScript warning in VSCode, but when I run `npm run compile`:

```bash
components/locale-provider/Localizable.tsx(11,1): error TS4060: Return type of exported function has or is using private name 'Localizable'.
components/upload/Upload.tsx(20,37): error TS4020: 'extends' clause of exported class 'Upload' has or is using private name 'UploadLocalizable'.
```

File: components/upload/Upload.tsx (hunk 1 above)
Comment (benjycui) [inline components/upload/Upload.tsx]: How?

**Comment by benjycui** [comment]
https://github.com/Microsoft/TypeScript/issues/14075

https://github.com/Microsoft/TypeScript/issues/14209

**Comment by benjycui** [comment]
Egg pained, close now, and waiting for `TypeScript@2.3`

File: components/upload/Upload.tsx (hunk 1 above)
Comment (bang88) [inline components/upload/Upload.tsx]: Maybe something like this.
```ts
export default class Table<T> extends Localizable<TableProps<T>, any>('Table', defaultLocale)
```

Comment (benjycui) [inline components/upload/Upload.tsx]: Why do developers need to customized Upload or other antd's components contextTypes?

Comment (bang88) [inline components/upload/Upload.tsx]: God know.
Never change antd's contextTypes.

Comment (benjycui) [inline components/upload/Upload.tsx]: Actually, Localizable works like mixin, so I don't understand what you are worrying about..

**Pull request closed without merging**